### PR TITLE
Update Documenter 0.27.23 => 1.2.1

### PR DESCRIPTION
**Pull request by mortenpi**
Currently mainly to test a few things on CI here, but will update this to an actual PR once 0.28.0 is out.

### CODE REVIEW

**Comment by mortenpi** [comment]
This should be good to go, except that #51328 should get merged first.

**Comment by LilithHafner** [comment]
CI failures look real

**Comment by mortenpi** [comment]
> CI failures look real

Yeah, they are. I think we do something weird with the Git repo on CI. Or, rather, I guess it's maybe not in a Git repo anymore once we build the docs.

cc @DilumAluthge @staticfloat

**Comment by staticfloat** [comment]
It is a git repo, it's just that [we download the built tarball and extract it into `julia-${shortcommit}`](https://github.com/JuliaCI/julia-buildkite/blob/main/pipelines/main/misc/doctest.yml#L25-L43).  I think you need to look for the stdlibs inside of `usr/share/julia/v1.X/stdlib`, since they won't be in `stdlib/foo-$hash`.

**Comment by mortenpi** [comment]
There are actually a few issues here. First, it looks like Git is not happy on the Windows builds:
```
┌ Error: Unable to determine the commit for the remote repository:
--
  | │ `git rev-parse --show-toplevel` failed
  | │   e.directory = "C:\\workdir"
  | │   exception =
  | │    Documenter.RepoCommitError("C:\\workdir", "`git rev-parse --show-toplevel` returned invalid path: /c/workdir", nothing)
```

https://buildkite.com/julialang/julia-master/builds/27823#018aa5df-6ef2-4de3-9477-9858a5119eaf/8109-8165

That's what prompted my comment about not being in a Git repo. But I think that's a Windows-only problem.

File: doc/make.jl
Comment (mortenpi) [inline doc/make.jl:96]: The second issue, that I initially missed, is that we can't find the stdlib:

```
ERROR: LoadError: Missing stdlib: /cache/build/default-amdci5-4/julialang/julia-master/stdlib/ArgTools-4eccde45ddc27e4f7fc9094b2861c684e062adb2
```

https://buildkite.com/julialang/julia-master/builds/27823#018aa5df-3848-40d1-9c9d-0623c89bdd21/713-752

@staticfloat So the `stdlib/` directory gets cleaned out, since we start anew, and just extract `usr`? Locally, `usr/share/julia/stdlib/v1.11/` are just symlinks to `stdlib/`. But the `.version` files are missing as far as I can tell.

Note: all this works locally

Comment (staticfloat) [reply doc/make.jl:96]: You're working in a build tree; but this is not a build tree, it's the result of `make install` into `usr`.

Comment (mortenpi) [reply doc/make.jl:96]: What's the best way to replicate the CI worktree locally? Should I just `cp -R` copy `usr/` into a clean Git worktree? The symlinks would still be there though?

It's a bit unfortunate that we're not building the docs in the build tree, to match what happens locally. But if there is a way to access the stdlib info consistently in both cases, then that should be fine.

Comment (staticfloat) [reply doc/make.jl:96]: There are no symlinks, because this is the result of `make install`, not just `make`.

You can take a clean `git` checkout, and then download the same tarball that is built from buildkite; this one for instance: https://buildkite.com/organizations/julialang/pipelines/julia-master/builds/27823/jobs/018aa5df-4d2c-41ad-a0be-226f905cdd6e/artifacts/018aa5e8-571c-41a6-9cb6-656215210e84

**Comment by LilithHafner** [comment]
Bump: https://github.com/JuliaLang/julia/issues/50035#issuecomment-1743216581

**Comment by mortenpi** [comment]
@LilithHafner So this is blocked by the CI vs local worktree differences, which I haven't had time to figure out, and not 100% sure when I'll have the bandwith. So on the off-chance you (or anyone) might have any time to help with that, I would greatly appreciate that.

This should be completely backportable though -- Documenter 1.0 works with Julia 1.6+. Although the `doc/Manifest.toml` may have to be manually re-resolved.

File: doc/make.jl
Comment (LilithHafner) [inline doc/make.jl:370]: I see you added a dep on `LibGit2` for `LibGit2.GITHUB_REGEX` here and Documenter.jl also [depends on LibGit2](https://github.com/JuliaDocs/Documenter.jl/blob/ae0188b2df02bc0e3f51f649832017abfccea3b3/Project.toml#L15). Is this info statement because Documentor also uses git via shelling out? If not it may be misleading. (somewhat related: https://github.com/JuliaLang/Pkg.jl/issues/2679)

**Review by LilithHafner — COMMENTED** [review]
CI is green; can we merge?

A few minor comments aside, the diff here LGTM. Unless you think that merging this will cause significant issues missed by CI, let's do this!

**Comment by mortenpi** [comment]
> CI is green; can we merge?

No, it needs a new Documenter version. It's depending on a branch right now. But soon -- it looks like between fixing the stdlib logic and switching Documenter to Git.jl, it is indeed finally working :tada:

File: doc/make.jl
Comment (mortenpi) [reply doc/make.jl:370]: This is just a debug print and I'll remove this. Generally, I need to do a pass on this PR before we merge, to clean it all up.

The issue I ran into here was that Git on the Windows runners is ancient (and I was trying to figure out what version it is exactly). But I resolved that by switching Documenter to Git.jl (which is the better than relying on the system's `git` binary anyway).

File: doc/make.jl
Comment (mortenpi) [reply doc/make.jl:313]: Correct. Technically, `strict` is gone, and the default is `warnonly = false`, which corresponds to `strict = true`.

File: doc/make.jl
Comment (mortenpi) [reply doc/make.jl:68]: It's in a script, so I don't think it makes any real difference. But I don't have an opinion either way :shrug:

File: doc/make.jl
Comment (mortenpi) [reply doc/make.jl:102]: This is very much specific to the Julia build. And not even the mail documentation build, but the stuff we do on CI. It should not be relevant for packages.

I wouldn't mind doing this somehow more cleanly in Documenter, but I don't think I have the energy to figure out what the solution should look like here.

Comment (LilithHafner) [reply doc/make.jl:102]: This is plenty clean as is if its a one-off for building Julia docs on CI.

File: doc/make.jl
Comment (mortenpi) [reply doc/make.jl]: Just a note here though: kinda silly to depend on LibGit2 just for this regex nowadays. https://github.com/JuliaDocs/Documenter.jl/issues/2363

**Comment by mortenpi** [comment]
Assuming all goes well on the CI, I think this is now ready to be merged.

**Review by LilithHafner — APPROVED** [review]

**Comment by N5N3** [comment]
Tried this locally and this PR seems to fix #44068.

**Comment by LilithHafner** [comment]
Oops, sorry I never got around to clicking merge once the CI passed. Thanks @N5N3!

**Comment by mortenpi** [comment]
It looks like this didn't fix the standard library links.. the CI build tree difference is causing yet more issues :cry:  (which I can replicate locally if I unpack a tarball into `usr/`).

Although, in this case, it also looks like Documenter is doing something a bit fishy, since it's failing to create the source links, but it's also not printing any warnings.

**Pull request merged**
